src/grafter_2/rdf/protocols.cljc

```diff
@@ -0,0 +1,616 @@
+
+(defrecord OffsetDate [date timezone])
+
+(def rdf:langString (->uri "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString"))
+
+;; TODO add tests to ensure that datatype-uri's etc are right
+;; everywhere we do string coercions.
+;;
+;; https://www.w3.org/TR/rdf11-new/#literals
+
+#?(:clj
+   (extend-type String
+     IRDFString
+     (lang [this]
+       nil)
+
+     IRawValue
+     (raw-value [this]
+       this)
+
+     IDatatypeURI
+     (datatype-uri [this]
+       gvx/xsd:string))
+
+   :cljs
+   (extend-type string
+     IRDFString
+     (lang [this]
+       nil)
+
+     IRawValue
+     (raw-value [this]
+       this)
+
+     IDatatypeURI
+     (datatype-uri [this]
+       gvx/xsd:string)))
+
+(defrecord LangString [string lang]
+  IRDFString
+  (lang [this]
+    (:lang this))
+
+  Object
+  (toString [this]
+    ;; TODO consider making this output the same as .toString on a RDF4j
+    ;; Literal.  Advantage is its more consistent with RDF4j etc... The
+    ;; disadvantage is that this implementation makes using str more intuitive
+    (:string this))
+
+  IRawValue
+  (raw-value [this]
+    (.toString this))
+
+  IDatatypeURI
+  (datatype-uri [this]
+    rdf:langString))
+
+#?(:cljs
+   ;; IEmptyableCollection fix for protocol bug seen in Chrome / Chromium
+   (extend-protocol IEmptyableCollection
+     LangString
+     (-empty [_] "")))
+
+(defn language
+  "Create an RDF langauge string out of a value string and a given
+  language tag.  Language tags should be keywords representing the
+  country code, e.g.
+
+  (language \"Bonsoir\" :fr)"
+  [s lang]
+  {:pre [(string? s)
+         lang
+         (keyword? lang)]}
+  (->LangString s lang))
+
+#?(:clj
+   (extend-type Literal
+     IRDFString
+     (lang [this]
+       (keyword (.orElse (.getLanguage this) nil)))
+
+     IRawValue
+     (raw-value [this]
+       (.stringValue this))
+
+     IDatatypeURI
+     (datatype-uri [this]
+       (URI. (str (.getDatatype this))))))
+
+(defrecord RDFLiteral [raw-value datatype-uri]
+  IRawValue
+  (raw-value [this]
+    (:raw-value this))
+
+  IDatatypeURI
+  (datatype-uri [this]
+    (:datatype-uri this))
+
+  IRDFString
+  (lang [this]
+    nil))
+
+(defn literal
+  "You can use this to declare an RDF typed literal value along with
+  its URI.  Note that there are implicit coercions already defined for
+  many core clojure/java datatypes, so for common datatypes you
+  shouldn't need this."
+  [val datatype-uri]
+  (->RDFLiteral (str val)
+                #?(:clj  (->java-uri datatype-uri)
+                   :cljs (->cljs-uri datatype-uri))))
+
+(extend-protocol IRawValue
+  nil
+  (raw-value [t]
+    t))
+
+#?(:clj
+   (extend-protocol IRawValue
+     Object
+     (raw-value [t]
+       t))
+   :cljs
+   (extend-protocol IRawValue
+     object
+     (raw-value [t]
+       t)))
+
+#?(:cljs
+   (extend-protocol IDatatypeURI
+     boolean
+     (datatype-uri [t] gvx/xsd:boolean)
+
+     string
+     (datatype-uri [t] gvx/xsd:string))
```

Comment:
Not sure about this `:cljs` IDatatypeURI extension. Should we leave this in?

This is vestigial code left over from when I had a "full" set of JS native type coercions, most notably for js/Number which could be an `xsd:integer`, `xsd:decimal` or `xsd:double`. We decided not to do this Number translation in case the coercion lost accurate information by being mistaken.